### PR TITLE
Another set of intermediate improvements

### PR DESCRIPTION
**Pull request by HechtiDerLachs**
This is from @simonbrandhorst 's and my work on the fibration hoppings.

### CODE REVIEW

File: src/AlgebraicGeometry/Schemes/ProjectiveVariety/Objects/Types.jl [hunk 1]
```diff
@@ -2,6 +2,7 @@
   X::ProjectiveAlgebraicSet{BaseRing,GradedRingType}
 
   function ProjectiveVariety(X::ProjectiveScheme{S, T}; check::Bool=true) where {S, T}
+    @check
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/ProjectiveVariety/Objects/Types.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/ProjectiveVariety/Objects/Types.jl [hunk 2]
```diff
@@ -16,6 +17,7 @@
   end
 
   function ProjectiveVariety(X::ProjectiveAlgebraicSet{S, T}; check::Bool=true) where {S, T}
+    @check
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/ProjectiveVariety/Objects/Types.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl [hunk 3]
```diff
@@ -51,6 +51,7 @@ function restrict(f::SpecMor, U::SpecOpen, V::SpecOpen; check::Bool=true)
     issubset(U, domain(f)) || error("$U is not contained in the domain of $f")
     issubset(V, codomain(f)) || error("$V is not contained in the codomain of $f")
     all(x->issubset(preimage(f, x), U), affine_patches(V)) || error("preimage of $V is not contained in $U")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl [hunk 4]
```diff
@@ -65,6 +66,7 @@ function restrict(
     issubset(U, domain(f)) || error("the given open is not an open subset of the domain of the map")
     issubset(V, codomain(f)) || error("the given open is not an open subset of the codomain of the map")
     issubset(preimage(f,V), U) || error("f(U) is not contained in V")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl [hunk 5]
```diff
@@ -77,6 +79,7 @@ function restrict(f::SpecOpenMor, W::AbsSpec, Y::AbsSpec; check::Bool=true)
   @check begin
     issubset(W, domain(f)) || error("$U is not contained in the domain of $f")
     issubset(W, preimage(f, Y)) || error("image of $W is not contained in $Y")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Constructors.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Types.jl [hunk 6]
```diff
@@ -51,6 +51,7 @@ mutable struct SpecOpenMor{DomainType<:SpecOpen,
       for g in f
         is_empty(subscheme(domain(g), pullback(g).(complement_equations(V)))) || error("image is not contained in the codomain")
       end
+      true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Morphisms/Types.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 7]
```diff
@@ -215,6 +215,7 @@ function restriction_map(
   @check begin
     X == hypersurface_complement(Y, h) || error("$X is not the hypersurface complement of $h in the ambient variety of $U")
     issubset(X, U) || error("$X is not a subset of $U")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 8]
```diff
@@ -305,6 +306,7 @@ function restriction_map(U::SpecOpen{<:AbsSpec{<:Ring, <:AbsLocalizedRing}},
   @check begin
     issubset(X, Y) || error("$X is not contained in the ambient scheme of $U")
     issubset(X, U) || error("$X is not a subset of $U")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 9]
```diff
@@ -329,6 +331,7 @@ function restriction_map(U::SpecOpen{<:AbsSpec{<:Ring, <:MPolyQuoRing}},
...
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 10]
```diff
@@ -344,6 +347,7 @@ function restriction_map(U::SpecOpen{<:AbsSpec{<:Ring, <:MPolyRing}},
...
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl [hunk 11]
```diff
@@ -420,6 +424,7 @@ function canonical_isomorphism(S::SpecOpenRing, T::SpecOpenRing; check::Bool=tru
   R == ambient_coordinate_ring(Y) || error("rings can not be canonically compared")
   @check begin
     (domain(S) == domain(T)) || error("open domains are not isomorphic")
+    true
```
Comment (HechtiDerLachs) [inline src/AlgebraicGeometry/Schemes/SpecOpen/Rings/Methods.jl]:
```suggestion
```

File: src/Rings/MPolyMap/MPolyQuo.jl [hunk 12]
```diff
@@ -83,6 +83,7 @@ function hom(R::MPolyQuoRing, S::NCRing, coeff_map, images::Vector; check::Bool
   @check begin
     _check_imgs_quo(R, S, imgs, coeff_map)
     _check_homo(S, imgs)
+    true
```
Comment (HechtiDerLachs) [inline src/Rings/MPolyMap/MPolyQuo.jl]:
```suggestion
```

File: src/Rings/MPolyMap/MPolyRing.jl [hunk 13]
```diff
@@ -107,6 +107,7 @@ function hom(R::MPolyRing, S::NCRing, coeff_map, images::Vector; check::Bool = t
   @check begin
     _check_imgs(S, imgs)
     _check_homo(S, imgs) # defined in MPolyAnyMap.jl
+    true
```
Comment (HechtiDerLachs) [inline src/Rings/MPolyMap/MPolyRing.jl]:
```suggestion
```

File: src/Rings/MPolyMap/MPolyRing.jl [hunk 14]
```diff
@@ -119,6 +120,7 @@ function hom(R::MPolyRing, S::NCRing, images::Vector; check::Bool = true)
...
```
Comment (HechtiDerLachs) [inline src/Rings/MPolyMap/MPolyRing.jl]:
```suggestion
```

File: src/Rings/mpoly-localizations.jl [hunk 15]
```diff
@@ -2631,6 +2631,7 @@ mutable struct MPolyLocalizedRingHom{
       for f in U
         is_unit(S(res(f))) || error("image of $f is not a unit in the codomain")
       end
+      true
```
Comment (HechtiDerLachs) [inline src/Rings/mpoly-localizations.jl]:
```suggestion
```

File: src/Rings/mpolyquo-localizations.jl [hunk 16]
```diff
@@ -984,6 +984,7 @@ constructor takes as input the triple
       for g in gens(modulus(underlying_quotient(L)))
         iszero(S(res(g))) || error("map is not well defined")
       end
+      true
```
Comment (HechtiDerLachs) [inline src/Rings/mpolyquo-localizations.jl]:
```suggestion
```

File: experimental/Schemes/WeilDivisor.jl [hunk 17]
```diff
@@ -49,6 +50,7 @@ stored as a formal linear combination over some ring ``R`` of
         isprime(D) || error("components of a divisor must be sheaves of prime ideals")
         dim(X) - dim(D) == 1 || error("components of a divisor must be of codimension one")
       end
+      true # Return true here for the check macro as all internal checks have passed.
```
Comment (HechtiDerLachs) [inline experimental/Schemes/WeilDivisor.jl]:
```suggestion
```

File: experimental/Schemes/WeilDivisor.jl [hunk 18]
```diff
@@ -38,6 +38,7 @@ stored as a formal linear combination over some ring ``R`` of
...
```
Comment (HechtiDerLachs) [inline experimental/Schemes/WeilDivisor.jl]:
```suggestion
```

File: experimental/Schemes/Types.jl [hunk 19]
```diff
@@ -551,6 +551,7 @@ identifications given by the glueings in the `default_covering`.
           end
         end
       end
+      true
```
Comment (HechtiDerLachs) [inline experimental/Schemes/Types.jl]:
```suggestion
```

File: experimental/Schemes/CoveredProjectiveSchemes.jl [hunk 20]
```diff
@@ -319,6 +333,19 @@ function blow_up_chart(W::AbsSpec{<:Field, <:MPolyRing}, I::MPolyIdeal;
...
+    # Cache the isomorphism on the complement of the center
+    p_res_dict = IdDict{AbsSpec, AbsSpecMor}()
+    for i in 1:ngens(I)
+      UW = PrincipalOpenSubset(W, gens(I)[i])
+      V = affine_charts(Y)[i]
+      VW = PrincipalOpenSubset(V, E(V))
+      p_res_dict[VW] = restrict(p_cov[V], VW, UW, check=false)
+      g = OO(UW).(gens(I))
+      set_attribute!(p_res_dict[VW], :inverse, 
+                     SpecMor(UW, VW, vcat([g[j]*inv(g[i]) for j in 1:ngens(I) if j != i], gens(OO(UW))))
```
Comment (simonbrandhorst) [inline experimental/Schemes/CoveredProjectiveSchemes.jl]:
```suggestion
                     SpecMor(UW, VW, vcat([g[j]*inv(g[i]) for j in 1:ngens(I) if j != i], gens(OO(UW))), check=false)
```

File: experimental/Schemes/CoveredProjectiveSchemes.jl [hunk 21]
```diff
@@ -351,6 +378,20 @@ function blow_up_chart(W::AbsSpec{<:Field, <:RingType}, I::Ideal;
...
+  # Cache the isomorphism on the complement of the center
+  p_res_dict = IdDict{AbsSpec, AbsSpecMor}()
+  for i in 1:ngens(I)
+    UW = PrincipalOpenSubset(W, gens(I)[i])
+    V = affine_charts(Y)[i]
+    VW = PrincipalOpenSubset(V, E(V))
+    p_res_dict[VW] = restrict(p_cov[V], VW, UW, check=false)
+    g = OO(UW).(gens(I))
+    set_attribute!(p_res_dict[VW], :inverse, 
+                   SpecMor(UW, VW, vcat([g[j]*inv(g[i]) for j in 1:ngens(I) if j != i], gens(OO(UW))))
```
Comment (simonbrandhorst) [inline experimental/Schemes/CoveredProjectiveSchemes.jl]:
```suggestion
                   SpecMor(UW, VW, vcat([g[j]*inv(g[i]) for j in 1:ngens(I) if j != i], gens(OO(UW))), check=false)
```

**Review by simonbrandhorst — APPROVED** [review]
Thank you looks good to me, once the tests pass.

File: experimental/Schemes/BlowupMorphism.jl [hunk 22]
```diff
@@ -475,3 +475,32 @@ function show_details(io::IO, Bl::BlowupMorphism)
...
+  inherit_glueings!(cod_cov, default_covering(Y))
+  XU = CoveredScheme(dom_cov)
+  YV = CoveredScheme(cod_cov)
+  p_res_cov = CoveringMorphism(dom_cov, cod_cov, iso_dict)
+  p_res = CoveredSchemeMorphism(XU, YV, p_res_cov)
+
+  # Assemble the inverse
+  iso_inv_dict = IdDict{AbsSpec, AbsSpecMor}()
+  for U in keys(iso_dict)
+    q = iso_dict[U]
```
Comment (fingolfin) [inline experimental/Schemes/BlowupMorphism.jl]:
```suggestion
  for (U,q) in iso_dict
```

File: experimental/Schemes/CoveredProjectiveSchemes.jl [hunk 23]
```diff
@@ -290,6 +290,20 @@ function blow_up_chart(W::AbsSpec{<:Field, <:MPolyRing}, I::MPolyIdeal;
     E = oscar.EffectiveCartierDivisor(Y, ID, trivializing_covering=domain(p_cov), check=false)
     set_attribute!(Y, :exceptional_divisor, E)
     set_attribute!(IPY, :exceptional_divisor, E)
+
+    # Cache the isomorphism on the complement of the center
+    p_res_dict = IdDict{AbsSpec, AbsSpecMor}()
+    for i in 1:ngens(I)
+      UW = PrincipalOpenSubset(W, gens(I)[i])
```
Comment (fingolfin) [inline experimental/Schemes/CoveredProjectiveSchemes.jl]: Don't do `gens(I)[i]`, we just spent a lot of effort to get rid of these things
```suggestion
      UW = PrincipalOpenSubset(W, gen(I,i))
```
or
```suggestion
      UW = PrincipalOpenSubset(W, I[i])
```
or for that matter, it seems you set `g=gens(I)` above, so
```suggestion
      UW = PrincipalOpenSubset(W, g[i])
```

File: experimental/Schemes/CoveredProjectiveSchemes.jl [hunk 24]
```diff
@@ -290,6 +290,20 @@ function blow_up_chart(W::AbsSpec{<:Field, <:MPolyRing}, I::MPolyIdeal;
...
+      V = affine_charts(Y)[i]
+      VW = PrincipalOpenSubset(V, E(V))
+      p_res_dict[VW] = restrict(p_cov[V], VW, UW, check=false)
+      g = OO(UW).(gens(I))
```
Comment (fingolfin) [inline experimental/Schemes/CoveredProjectiveSchemes.jl:301]:
```suggestion
      g = OO(UW).(g)
```

File: experimental/Schemes/CoveredProjectiveSchemes.jl [hunk 25]
```diff
@@ -319,6 +333,19 @@ function blow_up_chart(W::AbsSpec{<:Field, <:MPolyRing}, I::MPolyIdeal;
     E = oscar.EffectiveCartierDivisor(Y, ID, trivializing_covering=domain(p_cov), check=false) 
     set_attribute!(Y, :exceptional_divisor, E)
     set_attribute!(IPY, :exceptional_divisor, E)
+    # Cache the isomorphism on the complement of the center
+    p_res_dict = IdDict{AbsSpec, AbsSpecMor}()
+    for i in 1:ngens(I)
+      UW = PrincipalOpenSubset(W, gens(I)[i])
```
Comment (fingolfin) [inline experimental/Schemes/CoveredProjectiveSchemes.jl]: ditto

File: experimental/Schemes/CoveredProjectiveSchemes.jl [hunk 26]
```diff
@@ -351,6 +378,20 @@ function blow_up_chart(W::AbsSpec{<:Field, <:RingType}, I::Ideal;
   E = oscar.EffectiveCartierDivisor(Y, ID, trivializing_covering=domain(p_cov), check=false)
   set_attribute!(Y, :exceptional_divisor, E)
   set_attribute!(Bl_W, :exceptional_divisor, E)
+  
+  # Cache the isomorphism on the complement of the center
+  p_res_dict = IdDict{AbsSpec, AbsSpecMor}()
+  for i in 1:ngens(I)
+    UW = PrincipalOpenSubset(W, gens(I)[i])
```
Comment (fingolfin) [inline experimental/Schemes/CoveredProjectiveSchemes.jl]: ditto

File: experimental/Schemes/IdealSheaves.jl [hunk 27]
```diff
@@ -730,6 +740,7 @@ function match_on_intersections(
     end
 
 ## make sure we are working on consistent data
+    @check
     if check
```
Comment (fingolfin) [inline experimental/Schemes/IdealSheaves.jl]: Huh, shouldn't this be
```suggestion
    @check begin
```
(Genuine question, I forgot how we did it in the end, guess I should look at the `@check` docu...)

File: experimental/Schemes/WeilDivisor.jl [hunk 28]
```diff
@@ -300,6 +298,13 @@ generated by rational functions ``f₁,…,fᵣ ∈ K(X)``.
   end
 end
   
+@doc raw"""
+    linear_system(f::Vector, D::WeilDivisor; check::Bool=true)
+
+Return the linear system ``L`` generated by rational functions ``f₁,…,fᵣ ∈ K(X)``
```
Comment (fingolfin) [inline experimental/Schemes/WeilDivisor.jl:304]: Quite some unicode in this docstring...

File: src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl [hunk 29]
```diff
@@ -263,3 +263,18 @@ end
 
 morphism_type(X::AbsSpec, Y::AbsSpec) = morphism_type(typeof(X), typeof(Y))
 
+@doc raw"""
+    isomorphism_on_open_subsets(f::AbsSpecMor)
+
+For a birational morphism ``f : X → Y`` of `AbsSpec`s this 
+returns an isomorphism of affine schemes ``fᵣₑₛ : U → V`` which is
```
Comment (fingolfin) [inline src/AlgebraicGeometry/Schemes/AffineSchemes/Morphisms/Attributes.jl]: The `fᵣₑₛ` looks rather horribly over here. Also: Unicode....

File: src/Rings/MPolyMap/MPolyQuo.jl [hunk 30]
```diff
@@ -137,6 +137,7 @@ end
 
 # The two main evaluation methods
 function (F::MPolyAnyMap{<: MPolyQuoRing})(g)
+  @req parent(g) === domain(F) "g not in domain"
```
Comment (HechtiDerLachs) [inline src/Rings/MPolyMap/MPolyQuo.jl]: @simonbrandhorst: Did you do this? This is wrong!

Comment (simonbrandhorst) [reply src/Rings/MPolyMap/MPolyQuo.jl]: Yet I did it. I thought I removed it. Maybe it comes from an intermediate commit?

Comment (HechtiDerLachs) [reply src/Rings/MPolyMap/MPolyQuo.jl]: Who knows. But then at least we agree that this can be removed.

**Comment by simonbrandhorst** [comment]
Doctests fail.

File: experimental/Schemes/IdealSheaves.jl (hunk 27 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/IdealSheaves.jl]: Yes. Corrected.

File: experimental/Schemes/BlowupMorphism.jl (hunk 22 above)
Comment (HechtiDerLachs) [reply experimental/Schemes/BlowupMorphism.jl]: Thanks for the hint! I didn't know that.

**Pull request merged**